certbot/certbot/interfaces.py

```diff
@@ -70,35 +262,22 @@ class IPluginFactory(zope.interface.Interface):
 
     """
 
-    description = zope.interface.Attribute("Short plugin description")
-
-    def __call__(config, name):  # pylint: disable=signature-differs
-        """Create new `IPlugin`.
-
-        :param IConfig config: Configuration.
-        :param str name: Unique plugin name.
-
-        """
+    description: str = NotImplemented
+    """Short plugin description"""
 
-    def inject_parser_options(parser, name):
-        """Inject argument parser options (flags).
-
-        1. Be nice and prepend all options and destinations with
-        `~.common.option_namespace` and `~common.dest_namespace`.
-
-        2. Inject options (flags) only. Positional arguments are not
-        allowed, as this would break the CLI.
+    @abstractmethod
```

Comment:
This is essentially the ABC equivalent of the `IPluginFactory` interface. It represents the contract to create a new plugin object as used by `PluginEntryPoint.init` method in `certbot._internal.plugins.disco` module.

Could you elaborate on why we should remove it?

Reply:
I should have been clearer. What I think is a little awkward is defining this method as abstract which requires `plugins.common.Plugin` to give it the following definition to make it a concrete class.
```
    def __init__(self, config, name):  # pylint: disable=useless-super-delegation
        super().__init__(config, name)
```
I was wondering if instead of doing this, we could just remove the `@abstractmethod` line here. I believe this would give all `interfaces.Plugin` objects a concrete definition of `__init__` that sets the attributes we want on the class object and in cases like `plugins.common.Plugin`, they wouldn't need to override it anymore.

Reply:
Ok I see. I think it depends on what is the contract we want here. Either it is about how the object is created (so only the constructor, as an abstract method), or it is about how the object is created + the fact it has the parameters `name` and `config`.

My current implementation is about the second contract, trying to be as close as possible to `IPluginFactory` that was asserting this contract. But I think I went halfway on that matter and it is not very convincing: in any way, the `__init__` implementation can be overridden and the contract is broken.

So I think it depends on what we want:
* either it is about only the constructor, and we do exactly like you said
* or it is also about these two properties, and in this case, I think we should declare the abstract properties `name()` and `config()` on the `Plugin` interface, implementing them on `plugins.common.Plugin` by returning the private parameters `_name` and `_config`, then implement the constructor on this class by assigning to these two private parameters the values of the variables given to the constructor.

What do you think?

Reply:
I think the current interface should only be about the constructor. I think that's what the old one was and I don't see us assuming anywhere in Certbot that plugins have `config` or `name` attributes other than inside the plugin code. (If you try to verify this, be careful because some of the plugin discovery code uses a different `name` attribute referring to `pkg_resources.EntryPoint.name`.)

With that in mind, I personally like the first option better with one final tweak. Since we don't care what plugins do with the arguments given to `__init__`, I think the method here should be abstract with no implementation and we should instead continue to set `self.confg` and `self.name` in ` certbot/certbot/plugins/common.py` as opposed to the interface.

Reply:
And that's it.